libs/payments/cart/src/lib/checkout.service.ts

```diff
@@ -2,41 +2,101 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+import { Injectable } from '@nestjs/common';
+import { EligibilityService } from '@fxa/payments/eligibility';
+import { PayPalManager, PaypalCustomerManager } from '@fxa/payments/paypal';
 import {
   StripeClient,
   StripeManager,
   StripeSubscription,
+  SubplatInterval,
+  TaxAddress,
 } from '@fxa/payments/stripe';
-import { PayPalManager, PaypalCustomerManager } from '@fxa/payments/paypal';
-import { Injectable } from '@nestjs/common';
-import { ResultCart } from './cart.types';
 import { CheckoutError, CheckoutPaymentError } from './checkout.error';
+import {
+  CartTotalMismatchError,
+  CartEligibilityMismatchError,
+} from './cart.error';
+import { CartManager } from './cart.manager';
+import { ResultCart } from './cart.types';
+import { handleEligibilityStatusMap } from './cart.utils';
+import { ContentfulService } from '@fxa/shared/contentful';
 
 @Injectable()
 export class CheckoutService {
   constructor(
     private stripeClient: StripeClient,
     private stripeManager: StripeManager,
     private paypalCustomerManager: PaypalCustomerManager,
-    private paypalManager: PayPalManager
+    private paypalManager: PayPalManager,
+    private cartManager: CartManager,
+    private eligibilityService: EligibilityService,
+    private contentfulService: ContentfulService
   ) {}
 
   async prePaySteps(cart: ResultCart) {
-    // TODO:
-    // - If uid not present, create user stub account and update cart
-    // - Check if customer exists, create via stripe manager and update cart
-    // - Make sure shipping address of customer is updated with cart taxAddress
-    // - Check if customer is eligible for product in cart
-    // - Check if amount to be charged matches amount in cart
-    // - Cancel incomplete subscriptions to price user is subscribing to
-
-    const { stripeCustomerId } = cart;
-    if (!stripeCustomerId) {
-      // TODO: this is a placeholder and won't be necessary once the above todo items are done
-      throw new Error("Stripe customer doesn't exist");
+    let customer;
+    const taxAddress = cart.taxAddress as any as TaxAddress;
+
+    // if stripeCustomerId does not exist:
+    // create stub and update cart
+    if (!cart.stripeCustomerId) {
+      customer = await this.stripeManager.createPlainCustomer({
+        taxAddress: taxAddress,
+      });
+
+      await this.cartManager.updateFreshCart(cart.id, cart.version, {
+        stripeCustomerId: customer.id,
+      });
+    } else {
+      customer = await this.stripeManager.fetchActiveCustomer(
+        cart.stripeCustomerId
+      );
+    }
+
+    // validate customer is eligible for product via eligibility service
+    // throws cart eligibility mismatch error if no match found
+    const eligibility = await this.eligibilityService.checkEligibility(
+      cart.interval as SubplatInterval,
```

Comment:
This typecast should be removed by fixing the typings on the cart itself.

Reply:
Agreed. Filed a new ticket as mentioned in [this comment](https://github.com/mozilla/fxa/pull/16924#discussion_r1607275393). 👍